core/server/worker/src/main/java/alluxio/worker/block/BlockRWLockPool.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+import alluxio.resource.ResourcePool;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Class for managing block read/write locks. After obtaining a lock with
+ * {@link ResourcePool#acquire()}, {@link ResourcePool#release(Object)} must be called when the
+ * thread is done using the client.
+ */
+@ThreadSafe
+public final class BlockRWLockPool extends ResourcePool<ClientRWLock> {
+  private static final Logger LOG = LoggerFactory.getLogger(BlockRWLockPool.class);
+
+  @VisibleForTesting
+  final AtomicBoolean mHasReachedFullCapacity;
+  private final AtomicInteger mRemainingPoolResources;
+
+  /**
+   * Creates a new client read/write lock pool.
+   * @param maxCapacity the max capacity of the pool
+   */
+  public BlockRWLockPool(int maxCapacity) {
+    super(maxCapacity);
+    mHasReachedFullCapacity = new AtomicBoolean(false);
+    mRemainingPoolResources = new AtomicInteger(maxCapacity);
+
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.WORKER_BLOCK_LOCK_POOL_REMAINING_RESOURCES_KEY.toString(),
+        this::getRemainingPoolResources
+    );
+  }
+
+  /**
+   * Get the number of remaining pool resources.
+   * @return the number of remaining resources
+   */
+  @VisibleForTesting
+  public long getRemainingPoolResources() {
+    return mRemainingPoolResources.get();
+  }
+
+  private void maybeLogPoolReachesFullCapacity() {
+    if (this.size() == mMaxCapacity
+        && !mHasReachedFullCapacity.get()
+        && mHasReachedFullCapacity.compareAndSet(false, true)
+    ) {
+      LOG.warn("Block lock manager client RW lock pool exhausted. Capacity size {}", mMaxCapacity);
+    }
+  }
+
+  @Override
+  public ClientRWLock acquire() {
```

Comment:
do we really need this override method?

Reply:
yes. 
this is 
**this**.acquire(WAIT_INDEFINITELY, null);
instead of 
**super**.acquire(WAIT_INDEFINITELY, null);

Reply:
I don't write much java, but I think this override is redundant, we could safely remove it, and this.acquire would still be invoked.  I think the bytecode of this invocation is `invokeDynamic` rather than `invokeStatic`

Reply:
I think beinan is right. If not implemented, `BlockRWLockPool` will inherit the parent behavior and invoke `this.acquire(time, unit)` implicitly. 

I never expected to discuss bytecode with you guys when I clicked on this PR LOL. I guess the quickest way to verify is to test in the UT.

Reply:
yeah i was wrong... sorry for the confusion.